### DISCUSSION TITLE
livewire:navigated fired multiple times

### DISCUSSION THREAD

**Discussion by polycode-nz** [livewire/livewire#7019] [1. Help]
With V3.0.6 the event seems to be fired an extra time for every navigation that happens.

It does fire onload now where 3.0.5 didn't.

In a component blade file I have some javascript I `@push('scripts)` to the layout. To test:

```
@push('scripts')
    <script>
        document.addEventListener('livewire:navigated', () => {
            console.log('push scripts livewire navigated fired');
        });
    </script>
@endpush
```

In the document head I have an app.js file:

`
document.addEventListener('livewire:navigated', () => {
    console.log('app.js - livewire navigated');
});
`

Console onload:

<img width="638" alt="image" src="https://github.com/livewire/livewire/assets/48459194/bc238e2a-f19a-4bcf-8a67-3e5f6e7c6d02">

Console after one click of a `wire:navigate` link:

<img width="638" alt="image" src="https://github.com/livewire/livewire/assets/48459194/24debeb4-d29c-40f9-aaec-ea05db345bba">

Console after second click of a `wire:navigate` link:

<img width="638" alt="image" src="https://github.com/livewire/livewire/assets/48459194/93076de6-0299-456d-a712-573156cf7795">

Console after third click of a `wire:navigate` link:

<img width="638" alt="image" src="https://github.com/livewire/livewire/assets/48459194/4bfbfb46-cdee-4d4e-896a-fe32599cb720">

If I clear the console at this point and click a `wire:navigate` link a fourth time, console is:

<img width="638" alt="image" src="https://github.com/livewire/livewire/assets/48459194/fe33e2fc-bedf-438f-b0f6-19cd6e4784a9">

So always get one from the app.js file in the head then the one in the template seems to fire additively.

**Comment by joshhanley** [comment]
@polycode-nz thanks for reporting! I just tested this, and the issue is that you're missing the `data-navigate-once` attribute on your script tags. See here for details https://livewire.laravel.com/docs/navigate#scripts-in-the-body-are-re-evaluated

Hope this helps!

**Reply by polycode-nz** [reply]
Hey @joshhanley thanks for replying.

I do need the script to be ran/re-evaluated (once) per page though.

I should have said I'd tried that.

I changed the code in the component blade to be:

```
@push('scripts')
    <script data-navigate-once>
        document.addEventListener('livewire:navigated', () => {
            console.log('push scripts livewire navigated fired');
        });
    </script>
@endpush
```

It didn't change the console output from the above. After a load, 3 clicks of the nav, clear console the 4th click only shows:

<img width="551" alt="image" src="https://github.com/livewire/livewire/assets/48459194/b4c6adeb-66a5-4196-9a07-69165f189361">

I've got around it so far by setting a var to false, true on first go then checking if it's true and not running.

Besides this I'm having a tough time with JS and navigate in V3! [The note](https://livewire.laravel.com/docs/navigate#dont-rely-on-domcontentloaded) about just swapping out the DOMContentLoaded doesn't seem to be cutting it.

I noticed the latest release fixed `livewire:navigated` not firing onload which has helped.

**Comment by polycode-nz** [comment]
@joshhanley 

The `data-navigate-once` does work. It doesn't when setting a js var with a blade/component variable.

This runs once as you expect:
```
@push('scripts')
    <script data-navigate-once>
        document.addEventListener('livewire:navigated', () => {
            console.log('push scripts livewire navigated fired');
        });
    </script>
@endpush
```

So does this:
```
@push('scripts')
    <script data-navigate-once>
        document.addEventListener('livewire:navigated', () => {
            var potato = 'mashed';
            console.log('push scripts livewire navigated fired');
        });
    </script>
@endpush
```


This however does the additive behaviour:
```
@push('scripts')
    <script data-navigate-once>
        document.addEventListener('livewire:navigated', () => {
            var route = '{{ $route }}';
            var potato = 'mashed';
            console.log('push scripts livewire navigated fired');
        });
    </script>
@endpush
 ```

Where `$route` is a value sent from the components render method.

**Comment by gazzoy** [comment]
+1

**Comment by danpalmieri** [comment]
I just splited the code in:
```html
<script>
    let data = @json($data);
</script>

<script type="text/javascript" data-navigate-once>
    document.addEventListener('livewire:navigated', function() {
         // use the js data variable here
    });
</script>
```
and now is working!
